manta-crypto/src/algebra.rs

```diff
@@ -55,8 +58,11 @@ pub trait HasGenerator<G, COM = ()>
 where
     G: Group<COM>,
 {
+    /// Generator Type
+    type Generator;
```

Comment:
I think it's strange for this trait to have the generic `G: Group` and not have `Generator` related to `G` in any way.  

Usually a generator is an element of the group that it generates, so that `G` and `Generator` would be the same type. The only exception I can think of at the moment is Lie groups, whose generators belong to Lie algebras and therefore do have a different type -- that's totally irrelevant here, but even in that case there's a clear relation between the type of the generator and the group it generates.

Anyway, this isn't a deal-breaker for me but I think it's worth asking if we want to say anything concrete about how `G` and the `Generator` type relate. Feel free to dismiss this as silly mathematical nit-picking :)

Reply:
> I think it's strange for this trait to have the generic `G: Group` and not have `Generator` related to `G` in any way.  
> 
> Usually a generator is an element of the group that it generates, so that `G` and `Generator` would be the same type. The only exception I can think of at the moment is Lie groups, whose generators belong to Lie algebras and therefore do have a different type -- that's totally irrelevant here, but even in that case there's a clear relation between the type of the generator and the group it generates.
> 
> Anyway, this isn't a deal-breaker for me but I think it's worth asking if we want to say anything concrete about how `G` and the `Generator` type relate. Feel free to dismiss this as silly mathematical nit-picking :)

Yeah, you're still thinking too high-level, definitely didn't mean for this to have anything to do with Lie groups. `G` and `Generator` should actually be the same mathematical type but they need not be the same Rust types. For our case, we want `Generator` to be the pre-computed table and `G` to be the single group element, so although they really do mean the same mathematically, they don't mean the same in the code and in the performance characteristics (all the "side-effects" that are unobservable from a purely (classical) algebraic point-of-view). And it might be strange to think about, but it actually is _this `trait`_ that "relates" these two types together. The existence of this `trait` means that there is something that makes `Generator` relevant for operations with `G`. Actually, the part that's more strange is `G: Group<COM>`, `G` should probably get no type constraint at all. But we still want to make this about groups so that the word generator still has meaning here.

> Doesn't it make sense at this point to impose the extra constraint S: Ring? Also, maybe add a default implementation for any G and S = Z.

In general we want to reduce the number of constraints that are not _100%_ necessary for interface safety (and removing foot-guns). In this case `S` does not appear in this interface (to talk about groups we don't need to talk about scalar multiplication) and also $\mathbb{Z}$ is not a good default since it's not used very often (or even at all?) in cryptography.